File: skills/property-test/examples/property-types.ts
/**
 * Los 5 tipos de propiedades — ejemplos ejecutables (TypeScript / fast-check)
 * Referencia generativa para /property-test. Semilla fija = reproducible.
 *
 *   npm i -D fast-check
 *   npx vitest run property-types.ts   (o jest)
 */
import fc from 'fast-check';
import { describe, it } from 'vitest';

const SEED = 42;
const cfg = { seed: SEED, numRuns: 100 };

// --- funciones de ejemplo bajo prueba ---
const serializeAmount = (n: number) => n.toFixed(2);
const parseAmount = (s: string) => parseFloat(s);
const normalizeTag = (s: string) => s.trim().toLowerCase();
const taxOf = (price: number) => Math.round(price * 0.19 * 100) / 100;
const discountedPrice = (price: number, discount: number) => Math.max(0, price - discount);
const fastSum = (xs: number[]) => xs.reduce((a, b) => a + b, 0);
const refSum = (xs: number[]) => xs.length ? xs[0] + refSum(xs.slice(1)) : 0;

describe('5 property types', () => {
  // 1. ROUND-TRIP: parse(serialize(x)) == x (con tolerancia de redondeo a 2 decimales)
  it('round-trip: parse(serialize(x)) ≈ x', () => {
    fc.assert(fc.property(fc.float({ min: 0, max: 1e6, noNaN: true }), (x) => {
      const round2 = Math.round(x * 100) / 100;
      return Math.abs(parseAmount(serializeAmount(round2)) - round2) < 1e-9;
    }), cfg);
  });

  // 2. IDEMPOTENCY: f(f(x)) == f(x)
  it('idempotency: normalize(normalize(s)) == normalize(s)', () => {
    fc.assert(fc.property(fc.string(), (s) =>
      normalizeTag(normalizeTag(s)) === normalizeTag(s)), cfg);
  });

  // 3. MONOTONICITY: a <= b  =>  f(a) <= f(b)
  it('monotonicity: price↑ ⇒ tax↑', () => {
    fc.assert(fc.property(
      fc.float({ min: 0, max: 1e4, noNaN: true }),
      fc.float({ min: 0, max: 1e4, noNaN: true }),
      (a, b) => (a <= b ? taxOf(a) <= taxOf(b) : true)), cfg);
  });

  // 4. INVARIANTS: postcondición siempre cierta (el precio con descuento nunca es negativo)
  it('invariant: discountedPrice >= 0', () => {
    fc.assert(fc.property(
      fc.float({ min: 0, max: 1e6, noNaN: true }),
      fc.float({ min: 0, max: 1e6, noNaN: true }),
      (p, d) => discountedPrice(p, d) >= 0), cfg);
  });

  // 5. ORACLE: impl optimizada == impl de referencia
  it('oracle: fastSum == refSum', () => {
    fc.assert(fc.property(fc.array(fc.integer({ min: -1e3, max: 1e3 }), { maxLength: 50 }),
      (xs) => fastSum(xs) === refSum(xs)), cfg);
  });
});

File: skills/property-test/examples/property_types.py
"""
Los 5 tipos de propiedades — ejemplos ejecutables (Python / hypothesis)
Referencia generativa para /property-test. @seed fijo = reproducible.

    pip install hypothesis pytest
    pytest property_types.py
"""
from hypothesis import given, seed, settings, strategies as st

SEED = 42


# --- funciones de ejemplo bajo prueba ---
def serialize_amount(n: float) -> str:
    return f"{n:.2f}"


def parse_amount(s: str) -> float:
    return float(s)


def normalize_tag(s: str) -> str:
    return s.strip().lower()


def tax_of(price: float) -> float:
    return round(price * 0.19, 2)


def discounted_price(price: float, discount: float) -> float:
    return max(0.0, price - discount)


def fast_sum(xs: list[int]) -> int:
    return sum(xs)


def ref_sum(xs: list[int]) -> int:
    total = 0
    for x in xs:
        total += x
    return total


# 1. ROUND-TRIP: parse(serialize(x)) ~= x
@seed(SEED)
@settings(max_examples=100)
@given(st.floats(min_value=0, max_value=1e6, allow_nan=False, allow_infinity=False))
def test_round_trip(x):
    r = round(x, 2)
    assert abs(parse_amount(serialize_amount(r)) - r) < 1e-9


# 2. IDEMPOTENCY: f(f(x)) == f(x)
@seed(SEED)
@given(st.text())
def test_idempotency(s):
    assert normalize_tag(normalize_tag(s)) == normalize_tag(s)


# 3. MONOTONICITY: a <= b => f(a) <= f(b)
@seed(SEED)
@given(st.floats(0, 1e4, allow_nan=False), st.floats(0, 1e4, allow_nan=False))
def test_monotonicity(a, b):
    if a <= b:
        assert tax_of(a) <= tax_of(b)


# 4. INVARIANTS: postcondición siempre cierta
@seed(SEED)
@given(st.floats(0, 1e6, allow_nan=False), st.floats(0, 1e6, allow_nan=False))
def test_invariant_non_negative(p, d):
    assert discounted_price(p, d) >= 0


# 5. ORACLE: impl optimizada == impl de referencia
@seed(SEED)
@given(st.lists(st.integers(-1000, 1000), max_size=50))
def test_oracle(xs):
    assert fast_sum(xs) == ref_sum(xs)
